[feat] 포인트 기능 구현 및 결합

junn34 opened the pull request:
## ✅ 연관된 이슈
#21

## 📝 작업 내용
포인트 현금화 신청 기능 및 내역 조회 기능 구현했습니다.
포인트 적립 기능 구현했습니다.
기능들 하나로 모아서 #21 로 합쳤습니다.

## 🖼️ 스크린샷 (선택)

junn34 committed:
feat: 포인트 환전 신청 API 구현

junn34 committed:
feat: 포인트 환전 신청 API 구현

junn34 committed:
신청 내역 생성 및 조회 기API 구현

junn34 committed:
포인트 차감 추가

junn34 committed:
ㅕã보안 취약성 보완 및 변ìcamelcase로 변수이름 변경

junn34 committed:
feat/#12 pull 및 포인트 차감 suggestion ã반영

junn34 committed:
feat/#12 pull 및 포인트 차감 suggestion ㅘ 반영1

junn34 committed:
#21 브랜치 추가

junn34 committed:
적ë포인트 ½ 구현구현

junn34 committed:
수정1

junn34 committed:
review 반영

junn34 committed:
review 반영

junn34 commented:
/gemini review

junn34 committed:
리뷰 수정2

junn34 commented:
/gemini review

junn34 committed:
예외 처리 수정

junn34 committed:
record 타입 으로 dto 변경

junn34 commented:
/gemini review

junn34 committed:
[refactor] ë¹dto 변경으로 인한 코드 ë³수정

junn34 commented:
/gemini review

junn34 committed:
[refactor] postman 테스트 오류 해결중

junn34 committed:
postman 확인 완료

junn34 commented:
/gemini review

junn34 committed:
develop 브랜치와 ë충돌 해결

junn34 commented:
/gemini review

junn34 committed:
[refactor] 충돌 해결

junn34 commented:
/gemini review

junn34 committed:
[refactor] develop브랜치 pull

junn34 commented:
/gemini review

junn34 committed:
[refactor] postman 테스트중

junn34 committed:
[refactor] postman 테스트로 회원가입 및 로ê¸그인, 신현금화 신청, 신청 내역 조회, 포인트 적립 및 차감까지 성공

junn34 commented:
/gemini review

junn34 commented:
/gemini review

junn34 committed:
í[refactor] 퐁ë포인트 현금화 신청 승인부분 admin 역하할로 변경 및  전체적인 코드 리팩토링

junn34 committed:
[refactor] ãì수ì 최종 수정

junn34 commented:
/gemini review

junn34 committed:
포인트 기능 review 반영í했습니다.

junn34 commented:
/gemini review

junn34 committed:
[refactor] 비관적락 ㅈ거용

junn34 commented:
/gemini review

junn34 committed:
[refactor] 최종 ì수정

junn34 commented:
/gemini review

junn34 commented on src/main/java/dev/woori/wooriLearn/config/security/CurrentUserResolver.java:
```diff
@@ -0,0 +1,30 @@
+        Users user = userRepository.findByUserId(username)
+                .orElseThrow(() -> new CommonException(ErrorCode.UNAUTHORIZED, "로그인이 필요합니다."));
+        return user.getId();
+    }
```
jwt에 userId 넣으면 토큰 유출 시 복호화될 때 보안적으로 위험해서 적용안했습니다. 지금 서비스에선 DB 조회가 큰 부담이 될거같지않아 토큰에 넣는 방식을 적용하지않았습니다.

junn34 commented on src/main/java/dev/woori/wooriLearn/domain/account/repository/PointsHistoryRepository.java:
```diff
@@ -0,0 +1,34 @@
+            + "AND p.type = :type " // type 조건 추가
+            + "AND (:status IS NULL OR p.status = :status) "
+            + "AND (:start IS NULL OR p.createdAt >= :start) "
+            + "AND (:end IS NULL OR p.createdAt <= :end)")
```
JOIN FETCH를 필수적으로 적용해야하는 상황이 아니기 때문에 수정안했습니다.(불필요하게 USER정보까지 가져올수있음)

junn34 commented on src/main/java/dev/woori/wooriLearn/domain/account/repository/AccountRepository.java:
```diff
@@ -1,6 +1,17 @@
+public interface AccountRepository extends JpaRepository<Account, Long> {
+
+    // 유저가 가진 계좌 목록
+    List<Account> findAllByUser_Id(Long userId);
```
Review 가 틀렸어요 . User_Id 로 찾는게 맞습니다

junn34 commented on src/main/java/dev/woori/wooriLearn/config/security/CurrentUserResolver.java:
```diff
@@ -0,0 +1,30 @@
+        Users user = userRepository.findByUserId(username)
+                .orElseThrow(() -> new CommonException(ErrorCode.UNAUTHORIZED, "로그인이 필요합니다."));
+        return user.getId();
+    }
```
위의 review와 동일한 이유 ( user id 조회)

kimhm0828 reviewed (changes requested):
1. exception 처리
exception 처리는 에러 클래스를 따로 만들기보다 ErrorCode에 정의된 걸 쓰면 좋을 것 같습니다. 좀 더 명확하게 하고 싶으시다면 ErrorCode에 enum 값으로 추가해 주세요. 
Forbidden 같은 경우 그냥 `throw new CommonException(ErrorCode.Forbidden, "~~~")` 같은 식으로 작성하면 될 듯 하고
다른 에러들도... 굳이 Exception 클래스로까지 정의할 필요는 없을 것 같습니다.
출금 요청을 찾을 수 없다면 이것도 그냥 `throw new CommonException(ErrorCode.Entity_NOT_FOUND, "~~~")` 하면 되지 않나 싶어요.

----

2. CurrentUserResolver 관련

jwt 토큰으로 인증하는 흐름은
- 로그인 요청 시 jwt 토큰 발급 요청
- username / role / 생성 시간 / 만료 시간을 담은 토큰을 서버 내부의 비밀키로 서명해 생성
- 이후 Principal을 매개변수로 받는 요청이 들어오면 
내부적으로 요청에 포함된 jwt 토큰의 서명을 서버 내부의 비밀키로 다시 검증해 
현재 사용자에 대한 정보 추출

위와 같은 흐름으로 진행됩니다.

설명에서는 jwt 토큰에 username을 넣으면 복호화될 때 보안적으로 위험해질 수 있다고 하셨는데, 작성하신 코드(CurrentUserResolver)는 jwt 토큰 내부에 있는 username을 가져와 그걸 db에서 다시금 검증하는 로직이라 설명과 모순되는 부분이 있는 것 같습니다.
이미 토큰에 username을 넣고 principal은 그 안에 들어있는 username을 검증하는 형태인걸요...

CurrentUserResolver를 작성하신 이유는 아마 중요한 로직일 경우(ex. 포인트 환전 처리 등)
사용자가 정말로 실제 사용자가 맞는지 확인하기 위한 유틸리티성 클래스인 것 같은데
이마저도 jwt 토큰은 db에 존재하는 사용자의 로그인 요청에 의해 발급되기 때문에
jwt 토큰 내부에 있는 username은 반드시 db 내부에 존재할 수밖에 없습니다.

그래서 왜 이런 로직을 작성하신 건지 조금 이해하기 어렵습니다.
굳이 따지자면 사용자는 탈퇴했지만+토큰은 여전히 유효하고+그 토큰을 악의적인 사용자가 탈취해+중요한 로직에 접근하는 경우를 막기 위해서... 인 걸까요?

만약 사용하실 거라면 이 로직이 정말 보안상으로 중요한가?를 먼저 생각해 보시고
controller 아니면 service 로직 한 곳에서만 이용하는 방향으로 바꿔주시면 좋을 것 같습니다.
위치도 단순 db 조회만 하는 거라면 config 안이 아니라 domain/common 등의 위치로 옮겨주시는 게 좋을 것 같아요.

----


3. 메서드

`AccountRepository`에서
    `List<Account> findAllByUser_Id(Long userId);`
이 메서드는 사용하는 곳이 없는 것 같은데요...? 삭제하셔도 괜찮을 것 같습니다.

마찬가지로 `PointsHistoryRepository`의
`List<PointsHistory> findAllByUser_Id(Long userId);`
이 메서드도 사용하는 곳이 없는 것 같습니다.

----

4. `PointsHistoryRepository`의 `findByFilters`에서 
user 정보를 불필요하게 가져올 수 있어 굳이 join fetch를 적용하지 않는다고 하셨는데
`PointsExchangeService` 에서
```java
 List<PointsHistory> list = pointsHistoryRepository.findByFilters(
                userId, PointsHistoryType.WITHDRAW, statusEnum, start, end, sortOption
        );

        return list.stream()
                .map(h -> PointsExchangeResponseDto.builder()
                        .requestId(h.getId())
                        .userId(h.getUser().getId())
                        .exchangeAmount(h.getAmount())
                        .status(h.getStatus())
                        .requestDate(h.getCreatedAt())
                        .processedDate(h.getProcessedAt())
                        .message(mapStatusToMessage(h.getStatus()))
                        .build()
                ).toList();
```
위와 같이 `h.getUser().getId()`로 User 객체에 접근하고 있더라고요.
이러면 `PointsHistory`의 User 정보를 가져와야 해서 n+1 문제가 생길 수밖에 없습니다.
이 부분은 쿼리문 수정이 필요할 것 같아요.

----

5. `@PostMapping("")` 부분은 삭제 혹은 수정 처리 해주세요~
gemini review에 있던데 확인을 미처 못하신 것 같습니다.

----

+) git 작업 하실 때에는 그냥 파일 경로에 git bash 열어서 따로 작업하는 걸 추천드립니다...
저도 인텔리제이 터미널에서 해결하려고 했더니 한글만 입력하면 이상해지더라고요...

++) 구현한 기능에 대한 테스트 코드를 작성해 주시거나 postman 테스트 결과 스크린샷을 남겨주세요!

junn34 commented:
1. CommonException 사용하는 방식으로 수정했어요
2. controller한곳에서 사용하도록 변경했고 클래스이름도 PrincipalUtils로 명확하게 수정했습니다. domain/common/auth 패키지로 옮겼어요.
3. 안쓰는 메서드들 삭제했습니다.
4. 쿼리문 변경했습니다.

junn34 committed:
[refactor] 코드리뷰 반영

junn34 commented:
/gemini review

junn34 commented:
> 1. CommonException 사용하는 방식으로 수정했어요
> 2. controller한곳에서 사용하도록 변경했고 클래스이름도 PrincipalUtils로 명확하게 수정했습니다. domain/common/auth 패키지로 옮겼어요.
> 3. 안쓰는 메서드들 삭제했습니다.
> 4. 쿼리문 변경했습니다.
postman 테스트들 캡쳐는 노션에 api 명세에 페이지 만들었습니다. 성공사례만 캡쳐했고 실패 사례들도 잘되지만 예외적인 상황들은 테스트 코드로 해보려고합니다

junn34 commented on src/main/java/dev/woori/wooriLearn/domain/account/service/PointsExchangeService.java:
```diff
@@ -0,0 +1,185 @@
+            HistoryStatus status,
+            SortDirection sort
+    ) {
+        Long userId = userRepository.findByUserId(username)
+                .map(Users::getId)
+                .orElseThrow(() -> new CommonException(ErrorCode.ENTITY_NOT_FOUND, "사용자를 찾을 수 없습니다. userId=" + username));
```
username으로 직접받게되면 사용자의 존재 여부 구분 시 결국 조회를 2번하기 때문에 , 현재 코드 유지하는 것도 상관없다고 생각해서 review 적용안했습니다.

junn34 committed:
[refactor] gemini 리뷰 적용

junn34 commented:
/gemini review

junn34 committed:
[refactor] review 반영2

junn34 commented:
/gemini review

junn34 committed:
[chore] 문자열을 상수로 변환

junn34 commented:
/gemini review

junn34 committed:
[refactor] 테스트 코드 추가

junn34 commented:
/gemini review

junn34 committed:
[refactor] 테스트 코드 추가

junn34 commented:
/gemini review

junn34 committed:
[refactor] 포인트 환전신청부분 변경

junn34 commented:
/gemini review

junn34 committed:
[refactor] 코드 리뷰 반영

junn34 commented:
/gemini review

junn34 committed:
[refactor] 교착상태 해결

junn34 commented:
/gemini review

junn34 committed:
[refactor] temp txt파일 제거 및 review 반영

junn34 commented:
/gemini review

junn34 commented on src/main/java/dev/woori/wooriLearn/domain/account/service/PointsExchangeService.java:
```diff
@@ -0,0 +1,200 @@
+                ).toList();
+    }
+
+    // 출금 승인: Users → History 순서로 락
+    @Transactional
+    public PointsExchangeResponseDto approveExchange(Long requestId) {
+        // 비잠금으로 우선 조회
+        PointsHistory history = pointsHistoryRepository.findById(requestId)
+                .orElseThrow(() -> new CommonException(ErrorCode.ENTITY_NOT_FOUND, "출금 요청을 찾을 수 없습니다. requestId=" + requestId));
+
+        // Early check: not APPLY -> CONFLICT (avoid extra locks)
+        if (history.getStatus() != PointsStatus.APPLY) {
+            throw new CommonException(ErrorCode.CONFLICT, "이미 처리된 요청입니다.");
+        }
+
+        // Users 먼저 잠금 (lambda 캡처 변수는 final/유사-final 이어야 하므로 userId 보관)
+        Long userId = history.getUser().getId();
+        Users user = userRepository.findByIdForUpdate(userId)
+                .orElseThrow(() -> new CommonException(ErrorCode.ENTITY_NOT_FOUND, "사용자를 찾을 수 없습니다. Id=" + userId));
+
+        // 이후 History 잠금
+        history = pointsHistoryRepository.findAndLockById(requestId)
+                .orElseThrow(() -> new CommonException(ErrorCode.ENTITY_NOT_FOUND, "출금 요청을 찾을 수 없습니다. requestId=" + requestId));
+
+        if (history.getStatus() != PointsStatus.APPLY) {
+            throw new CommonException(ErrorCode.CONFLICT, "이미 처리된 요청입니다.");
+        }
```
이미 락 과정이 통일되어있고, fetch join 도입 시 다시 락을 별도의 쿼리로 가져와야해서 지금 코드 유지해도 된다고생각하여 반영하지 않았습니다.

kimhm0828 replied:
```java
@Lock(LockModeType.PESSIMISTIC_WRITE)
@Query("""
    SELECT ph FROM PointsHistory ph
    JOIN FETCH ph.user u
    WHERE ph.id = :userId
""")
Optional<PointsHistory> findWithUserForUpdate(@Param("userId") Long userId);
```
이런 식으로 DB 쿼리문을 바꾸면 한번에 두 엔티티에 대한 락을 설정할 수 있다고 합니다
한번 시도해 보셔도 좋을 것 같아요

kimhm0828 replied:
+) 락에 대한 타임아웃 로직도 추가해 주시면 좋을 것 같습니다
현재 코드는 타임아웃이 없어서 A 요청이 락을 잡고 있으면 B 요청은 계속 대기만 하게 돼요
시간지연이 너무 길어지면 잠시 후 다시 시도해 달라는 에러를 띄워도 괜찮을 것 같습니다

kimhm0828 commented on src/main/java/dev/woori/wooriLearn/domain/account/controller/AdminPointsExchangeController.java:
```diff
@@ -0,0 +1,30 @@
+
+    @PostMapping("/approve/{requestId}")
+    @PreAuthorize("hasRole('ADMIN')")
+    public ResponseEntity<BaseResponse<?>> approve(@PathVariable Long requestId) {
+        PointsExchangeResponseDto dto = pointsExchangeService.approveExchange(requestId);
+        return ApiResponse.success(SuccessCode.OK, dto);
+    }
```
`return ApiResponse.success(SuccessCode.OK, pointsExchangeService.approveExchange(requestId))`와 같은 형식으로 단순화할 수 있을 것 같습니다

kimhm0828 replied:
mapping도 Post가 아니라 Put으로 바꿔주세요

kimhm0828 commented on src/main/java/dev/woori/wooriLearn/domain/account/controller/PointsExchangeController.java:
```diff
@@ -0,0 +1,52 @@
+    ) {
+        String username = PrincipalUtils.requireUsername(principal);
+        PointsExchangeResponseDto response = pointsExchangeService.requestExchange(username, dto);
+        return ApiResponse.success(SuccessCode.CREATED, response);
```
여기도 마찬가지로 하나로 합칠 수 있을 것 같아요

kimhm0828 commented on src/main/java/dev/woori/wooriLearn/domain/account/controller/PointsExchangeController.java:
```diff
@@ -0,0 +1,52 @@
+            @RequestParam(required = false) String startDate,
+            @RequestParam(required = false) String endDate,
+            @RequestParam(required = false, defaultValue = "ALL") HistoryStatus status,
+            @RequestParam(required = false, defaultValue = "DESC") SortDirection sort
```
이 부분을 Request DTO로 묶을 수도 있을 것 같습니다

kimhm0828 commented on src/main/java/dev/woori/wooriLearn/domain/account/controller/PointsExchangeController.java:
```diff
@@ -0,0 +1,52 @@
+        String username = PrincipalUtils.requireUsername(principal);
+        List<PointsExchangeResponseDto> history = pointsExchangeService
+                .getHistory(username, startDate, endDate, status, sort);
+        return ApiResponse.success(SuccessCode.OK, history);
```
여기도 통합할 수 있을 것 같아요

junn34 replied:
위의 리뷰대로 requestdto 로 수정하면서 history 코드가 길어져서 따로 변수로 명시했습니다

kimhm0828 commented on src/main/java/dev/woori/wooriLearn/domain/account/dto/PointsDepositRequestDto.java:
```diff
@@ -0,0 +1,6 @@
+package dev.woori.wooriLearn.domain.account.dto;
+
+public record PointsDepositRequestDto(
+        @jakarta.validation.constraints.NotNull @jakarta.validation.constraints.Positive Integer amount,
+        String reason
```
이거 원래 이렇게 써지나요...? import한 다음 `@NotNull`로 쓸 수 있는 걸로 알고 있습니다

kimhm0828 commented on src/main/java/dev/woori/wooriLearn/domain/account/dto/PointsExchangeRequestDto.java:
```diff
@@ -0,0 +1,7 @@
+public record PointsExchangeRequestDto(
+        @jakarta.validation.constraints.NotNull @jakarta.validation.constraints.Positive Integer exchangeAmount,
+        @jakarta.validation.constraints.NotBlank String accountNum,
+        @jakarta.validation.constraints.NotBlank String bankCode
```
여기도 마찬가지로 바꿀 수 있으면 바꿔주세요

kimhm0828 commented on src/main/java/dev/woori/wooriLearn/domain/account/entity/PointsHistory.java:
```diff
@@ -24,10 +25,31 @@ public class PointsHistory {
+    private LocalDateTime processedAt;
+
+    @Column(name = "fail_reason")
+    private String failReason;
```
코드 확인해 보니까 이 부분을 상수로 관리하고 있던데
그러면 아예 상수들을 다 통합해서 enum으로 관리하는 것도 괜찮지 않을까 싶습니다

kimhm0828 commented on src/main/java/dev/woori/wooriLearn/domain/user/entity/Users.java:
```diff
@@ -24,7 +27,40 @@ public class Users extends BaseEntity {
+
+
+
+    public void addPoints(Integer amount) {
+        if (amount == null || amount <= 0) {
+            throw new CommonException(ErrorCode.INVALID_REQUEST, "포인트 증액은 양수여야 합니다.");
+        }
+        this.points += amount;
+    }
+
+    public void subtractPoints(Integer amount) {
+        if (amount == null || amount <= 0) {
+            throw new CommonException(ErrorCode.INVALID_REQUEST, "포인트 차감은 양수여야 합니다.");
+        }
+        if (this.points < amount) {
+            throw new CommonException(ErrorCode.CONFLICT, "포인트가 부족합니다.");
+        }
+        this.points -= amount;
+    }
```
에러를 던지게 하기보다는 boolean 값으로 처리하는 것도 방법일 것 같습니다
두 메서드를 합쳐서 calculatePoints 같은 메서드를 만들고
```java
public boolean calculatePoints(Integer amount){
    // 포인트가 마이너스일 경우 return false
    // else return true
}
```
같은 식으로 구현하고
PointsExchangeService 쪽의 try-catch 부분을 if-else로 변경하는 것도 방법일 것 같아요

junn34 replied:
calculatePoints로 만들어 로직을 하나로 합치면 의미가 모호해지고 , boolean으로 처리하면 롤백이 안되서 수정하지않았습니다.

kimhm0828 commented on src/main/java/dev/woori/wooriLearn/domain/account/service/PointsExchangeService.java:
```diff
@@ -0,0 +1,200 @@
+    // 출금 승인: Users → History 순서로 락
+    @Transactional
+    public PointsExchangeResponseDto approveExchange(Long requestId) {
+        // 비잠금으로 우선 조회
+        PointsHistory history = pointsHistoryRepository.findById(requestId)
+                .orElseThrow(() -> new CommonException(ErrorCode.ENTITY_NOT_FOUND, "출금 요청을 찾을 수 없습니다. requestId=" + requestId));
+
+        // Early check: not APPLY -> CONFLICT (avoid extra locks)
+        if (history.getStatus() != PointsStatus.APPLY) {
+            throw new CommonException(ErrorCode.CONFLICT, "이미 처리된 요청입니다.");
+        }
```
이 부분을 묶어서
```java
@Transactional(readOnly = true)
private PointsHistory validateRequest(Long requestId) {
    PointsHistory history = pointsHistoryRepository.findById(requestId)
        .orElseThrow(() -> new CommonException(ErrorCode.ENTITY_NOT_FOUND, "출금 요청을 찾을 수 없습니다. requestId=" + requestId));
    if (history.getStatus() != PointsStatus.APPLY) {
        throw new CommonException(ErrorCode.CONFLICT, "이미 처리된 요청입니다.");
    }
    return history;
}
```
이런 식으로 read-only로 분리하면 속도가 향상될 수 있다고 합니다

junn34 replied:
private으로 할 때는 트랜잭션이 바뀌지않아 성능 개선이 안된다고하네요.
분리했을 떄의 속도 체감(dirty checking 미활성화 등)은 발생하지 않는다고합다

kimhm0828 reviewed (changes requested):
리뷰 남겼습니다!
이것저것 많긴 한데... 좀 아닌 것 같다 싶으시면 반영하지 않으셔도 무방합니다🙇‍♀️
대신 이유는 꼭 작성해 주세요!

+) 혹시 PointsExchangeService랑 PointsDepositService로 나눈 이유가 있을까요?
하나로 합쳐도 괜찮을 것 같습니다.
Controller도 마찬가지로 하나로 합칠 수 있을 것 같아요.
(물론 이것도 이유가 있으시다면 반영하지 않으셔도 무방합니다!)

junn34 committed:
[refactor] 조회 기간 선택 추가 및 PL 리뷰반영

junn34 commented:
/gemini review

junn34 committed:
[refactor] test 코드 변경

junn34 commented:
/gemini review

junn34 committed:
[refactor] medium priority 수정

junn34 commented:
/gemini review

junn34 commented on src/main/java/dev/woori/wooriLearn/domain/account/controller/PointsExchangeController.java:
```diff
@@ -0,0 +1,58 @@
+                username,
+                request.getStartDate(),
+                request.getEndDate(),
+                request.getPeriod(),
+                request.getStatus(),
+                request.getSort()
+        );
+
+        return ApiResponse.success(SuccessCode.OK, history);
+    }
```
pl님의 방법이 더 보기 깔끔한거같아서 기존대로 dto 로 쓰겠습니다.

junn34 committed:
[refactor] n+1 문제 해결

junn34 commented:
/gemini review

junn34 committed:
[refactor] exception 변경

junn34 commented:
/gemini review

junn34 committed:
[refactor] 과도한  비관적 락 축소

junn34 commented:
/gemini review

junn34 committed:
[refactor] 다시 변경

junn34 commented:
/gemini review

junn34 committed:
[chore] enum 타입 명시 및 문구 변경 및 인자 변경

kimhm0828 commented on src/main/java/dev/woori/wooriLearn/domain/account/controller/AdminPointsExchangeController.java:
```diff
@@ -0,0 +1,49 @@
+            @RequestParam(defaultValue = "1") @Min(1) Integer page,
+            @RequestParam(defaultValue = "20") @Min(1) @Max(200) Integer size,
+            @RequestParam(required = false) Long userId
+    ) {
```
이 부분을 DTO로 묶어서 처리할 수 있을 것 같아요
```java
public record PointsHistoryReqDto(
    String startDate,
    String endDate,
    ....
){
}
```
이런 식으로 할 수 있지 않을까요?
gemini 리뷰도 이 메서드랑 일관성을 맞추라는 의미에서 나온 것 같습니다

junn34 replied:
기존에 있던 PointsHistorySearchRequestDto 재사용해서 수정했습니다.

kimhm0828 commented on src/main/java/dev/woori/wooriLearn/domain/common/auth/PrincipalUtils.java:
현재 jwt 토큰 생성 시
```java
  UsernamePasswordAuthenticationToken auth =
                        new UsernamePasswordAuthenticationToken(username, null,
                                List.of(new SimpleGrantedAuthority(role.name())));
```
위와 같은 형식으로 생성되고 있기 때문에
`Object principal`의 타입은 반드시 String이 됩니다.

따라서 굳이 PrincipalUtils를 쓸 필요 없이
Controller에서는 `@AuthenticationPrincipal String username`과 같은 식으로 바로 받을 수 있습니다.

kimhm0828 commented on src/main/java/dev/woori/wooriLearn/config/CustomUserDetails.java:
CustomUserDetails가 활용되는 곳이 있나요? 없다면 굳이 사용하지 않고 삭제해도 무방할 것 같습니다.

kimhm0828 commented on src/main/java/dev/woori/wooriLearn/domain/account/service/PointsExchangeService.java:
```diff
@@ -0,0 +1,284 @@
+        } catch (jakarta.persistence.LockTimeoutException
+                 | jakarta.persistence.PessimisticLockException
+                 | jakarta.persistence.QueryTimeoutException
+                 | org.springframework.dao.PessimisticLockingFailureException e) {
```
혹시 이부분 그냥 LockTimeoutException 같은 방식으로 작성할 수 있을까요?
안되는 거면 그대로 두셔도 무방합니다

kimhm0828 commented on src/main/java/dev/woori/wooriLearn/domain/user/repository/UserRepository.java:
```diff
@@ -1,12 +1,35 @@
+    @Query("SELECT u FROM Users u WHERE u.userId = :userId")
+    Optional<Users> findByUserIdForUpdate(@Param("userId") String userId);
+
+    @Lock(LockModeType.PESSIMISTIC_WRITE)
+    @QueryHints({
+            @QueryHint(name = "javax.persistence.lock.timeout", value = "2000"),
+            @QueryHint(name = "org.hibernate.timeout", value = "3")
+    })
+    @Query("SELECT u FROM Users u WHERE u.id = :id")
+    Optional<Users> findByIdForUpdate(@Param("id") Long id);
```
`@QueryHints`를 쓴 특별한 이유가 있을까요?
만약 시간제한을 걸어두고 싶었던 거라면 서비스 단에서 `@Transaction(timeout=3)` 같은 방식으로 설정할 수 있을 것 같습니다.

junn34 replied:
일정 수치만큼 대기하게 되면 빠르게 실패반응하도록 하기위해서 썼습니다.(무기한 대기 방지)

kimhm0828 replied:
굳이 `@QueryHints`로 쿼리 수준에서 시간제한을 걸 이유가 없을 것 같은데 어떻게 생각하시나요?

kimhm0828 reviewed (changes requested):
코멘트 남겼습니다~
이번에도 역시 전부 반영하실 필요는 없고 이유가 있으시다면 반영하지 않으셔도 무방합니다!

junn34 committed:
hello

junn34 committed:
[refactor] 리뷰 반영

junn34 commented:
hello라는 커밋은 잘못올렷습니다.

kimhm0828 commented on src/main/java/dev/woori/wooriLearn/domain/account/controller/AccountAuthController.java:
```diff
@@ -28,7 +28,7 @@ public class AccountAuthController {
 
     /**
-     * 인증번호(OTP) 발급/재발급 요청
+     * 인증번호(OTP) 발급/재발급 요청  s
```
오타 삭제해 주세요!

kimhm0828 reviewed (commented):
코멘트 남겼습니다~

junn34 commented:
@QueryHints 문 제거했습니다.

junn34 committed:
[refachore¬뷰 반영

kimhm0828 commented on src/main/java/dev/woori/wooriLearn/domain/account/controller/PointsDepositController.java:
```diff
@@ -0,0 +1,29 @@
+            @AuthenticationPrincipal String username,
+            @RequestBody PointsDepositRequestDto dto
+    ) {
+        return ApiResponse.success(SuccessCode.CREATED, pointsDepositService.depositPoints(username, dto));
```
`SuccessCode.OK`로 바꿔주시는 게 좋을 것 같습니다
CREATED는 회원가입과 같이 새로운 리소스가 생성되었을 때 사용하는 게 좋아요

kimhm0828 commented on src/main/java/dev/woori/wooriLearn/domain/account/service/PointsExchangeService.java:
```diff
@@ -0,0 +1,341 @@
+                pageResult.getTotalElements(),
+                pageResult.getTotalPages(),
+                pageResult.hasNext()
+        );
```
포인트 내역을 페이지로 응답해주는 코드가 사용자와 관리자에서 중복되는 것 같습니다
하나의 메서드로 묶을 수 있을 것 같아요

kimhm0828 reviewed (commented):
리뷰 남겼습니다~ 수정한 후에 테스트도 반드시 진행해 주세요

junn34 committed:
[refactor] 포인트 기능 공통 메서드로 통합

kimhm0828 committed:
[refactor] 코드 구조 개선 및 테스트 코드 추가

kimhm0828 committed:
[fix] 브랜치 충돌 문제 해결

# Conflicts:
#	src/main/java/dev/woori/wooriLearn/config/security/SecurityConfig.java
#	src/main/java/dev/woori/wooriLearn/domain/edubankapi/autopayment/entity/AutoPayment.java

kimhm0828 reviewed (approved):
수고하셨습니다~
코드 구조 일부 수정해서 커밋해 두었으니 확인해 주시고 머지해 주시면 됩니다!
